internal/pkg/service/common/etcdop/watch.go

```diff
@@ -75,56 +79,52 @@ func (v Prefix) GetAllAndWatch(ctx context.Context, client *etcd.Client, handleE
 		if err != nil {
 			// GetAll error is fatal
 			handleErr(err)
-			close(ch)
+			close(outCh)
+			close(initDoneCh)
+			return
 		}
 
 		// Continue with Watch where GetAll ended
+		close(initDoneCh)
 		opts = append(opts, etcd.WithRev(itr.Header().Revision+1))
-		v.doWatch(ctx, client, handleErr, ch, opts...)
+		v.doWatch(ctx, client, handleErr, outCh, opts...)
 	}()
 
-	return ch
+	return outCh, initDoneCh
 }
 
-func (v Prefix) doWatch(ctx context.Context, client etcd.Watcher, handleErr func(err error), ch chan Event, opts ...etcd.OpOption) {
+func (v Prefix) doWatch(ctx context.Context, client etcd.Watcher, handleErr func(err error), outCh chan Event, opts ...etcd.OpOption) {
 	opts = append([]etcd.OpOption{etcd.WithPrefix()}, opts...)
-	rawCh := client.Watch(ctx, v.Prefix(), opts...)
 	go func() {
+		defer close(outCh)
+
+		// In case of an error, the watch channel can be closed.
+		// It will be recreated, and it will continue from the last revision.
+		revision := int64(0)
 		for {
 			select {
 			case <-ctx.Done():
```

Comment:
Teda pribudol tu este jeden nadradeny `for`, aby to bolo prehladnejsie, tak som cast kodu presunul do `processWatchEvents`.